new_functions.py

```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
+    if "val_loss" in history:
+        ax.plot(history["val_loss"], color="orange", label="Validation loss")
+    ax.set_xlabel("Epoch")
+    ax.set_ylabel("Loss")
+    ax.set_title("Training History (Final Re-Fit)")
     ax.legend()
     return ax
+
+
+def plot_absolute_deviation_to_ax(
+    fig: Figure,
+    ax: Axes,
+    x_values: np.ndarray,
+    absolute_deviation: np.ndarray,
+    x_axis_label_td: str,
+    y_axis_label_td: str,
+    title: str = "",
+    height: float = 6,
+    width: float = 8,
+    label: str = "",
+) -> Axes:
+    """
+    Plot absolute deviation to ax
+
+    :param ax: Axes object to plot on
+    :param absolute_deviation: Absolute deviation in total density
+
+    :return: ax object with absolute deviations plotted
+    """
+    if len(absolute_deviation.shape) == 1:
+        ax.plot(x_values, absolute_deviation, linestyle="-", label=label)
+        ax.legend()
+    else:
+        for residuals in absolute_deviation:
+            ax.plot(x_values, residuals, color="k", linestyle="-", alpha=0.2)
+    ax.set_xlabel(x_axis_label_td)
+    ax.set_ylabel(y_axis_label_td)
+    fig.set_figheight(height)
+    fig.set_figwidth(width)
+    ax.set_ylim(bottom=0)
+    ax.set_title(title)
+    plt.tight_layout()
+    return fig, ax
+
+
+def add_noise_to_form_factor(
+    form_factor_df: pd.DataFrame,
+    steepness: float = 10.0,
+    max_noise_magnitude: float = 1.0,
+    probability_of_noise: float = 1.0,
+) -> pd.DataFrame:
+    """
+    Adds random noise to each row of the input DataFrame, with noise amplitude determined by a pure sigmoid function (scaled by `max_noise_magnitude`) across the columns. The amplitude starts near 0 in the first column and approaches `max_noise_magnitude` in the last column.
+
+    The amplitude for column i is calculated as:
+        amplitude(i) = max_noise_magnitude * [ 1 / (1 + exp(-steepness * (fraction - 0.5))) ]
+    where fraction = i / (num_cols - 1).
+
+    :param form_factor_df: A pandas DataFrame where each row is a form factor
+    :param steepness: Controls how quickly the sigmoid transitions from 0 to max_noise_magnitude. Higher values make the transition sharper around the midpoint.
+    :param max_noise_magnitude: The maximum amplitude of the noise (occurs at the highest q value).
+    :param probability_of_noise: The chance that a single form factor will be noisy.
+
+    :return: A new pandas DataFrame of the same shape, with column-wise noise added.
+    """
+    num_rows, num_cols = form_factor_df.shape
+
+    # Array of values from 0 to 1 in the dimension of q, to make the sigmoid
+    fraction_array = np.linspace(0.0, 1.0, num_cols)
+
+    # Noise amplitudes follow the logistic (sigmoid) function, scaled by max_noise_magnitude
+    # At fraction=0, amplitude ~ 0; at fraction=1, amplitude ~ max_noise_magnitude
+    noise_amplitudes = max_noise_magnitude * (
+        1.0 / (1.0 + np.exp(-steepness * (fraction_array - 0.5)))
+    )
```

Comment:
```suggestion
    # Noise amplitudes follow the logistic (sigmoid) function, scaled by max_noise_magnitude
    # At fraction=0, amplitude ~ 0; at fraction=1, amplitude ~ max_noise_magnitude
    fraction_array = np.linspace(0.0, 1.0, num_cols)
    noise_amplitudes = max_noise_magnitude * (
        1.0 / (1.0 + np.exp(-steepness * (fraction_array - 0.5)))
    )
```
(just a bit more compact. the linspace call doesn't need explanation)